### ISSUE TITLE
OpensearchUser,Role,RoleBinding resources are not re-applied after SecurityConfig reconciliation

### ISSUE THREAD

**Issue by max-frank** [opensearch-project/opensearch-k8s-operator#254]
As SecurityConfig reconciliation overwrites the internal users in `.opendistro_security` (if the field is set in the secret).
This is intended behaviour and to be expected, but as of now also means that all users (or roles and role bindings) created with the newly supported resources will also be removed and `.opendistro_security` will be set to just the values in the SecurityConfig secret. The current reconciliation logic is not able to detect this change and re-apply the already configured users, roles and role bindings. Meaning while the k8s resources will still be present and report 'unchanged' on apply, opensearch it self will not reflect the configuration correctly.

Now for roles and role bindings we can avoid this "problem" by simply not including them in the SecurityConfig secret and only configuring them using the resources, but for OpensearchUser resources this is not really possible due to limitations of the operator. As of right now the operator only supports Basic Auth for any processes (other than SecurityConfig reconciliation) involving a clusters OpenSearch API. Including applying OpensearchUser resources via the API. This means we always need at least one admin user configured via the security config so it can be used by the operator to manage the cluster. The only possible work around I can see for now would be to never ever modify the internal_users.yaml entry or removed it from the SecurityConfig secret after initial deploy.

It might be worth adding the above problem as disclaimer in the documentation for now and if possible implement logic in the controller to re-apply the resources on SecurityConfig reconciliation/detect that the configuration has drifted. Doing this would at least eventually restore the users.

Also ideally it would be possible to configure the cluster deployment to not rely on basic auth, but instead use something like client certificates instead. As is also mentioned in https://github.com/Opster/opensearch-k8s-operator/issues/221

**Comment by swoehrl-mw** [comment]
Hi @max-frank. You raise an interesting situation. The securityconfig is needed initially to configure the user the operator then uses to provision any other users. But changing the securityconfig overwrites the other users.
Your suggestion of reapplying the Users after a securityconfig change might work but I think its not a good solution as then there is a time window between securityconfig applied and Users recreated where these users have no access.

I see two possibilities:
* Switching the operator to use a client certificate when provisioning User/Role/RoleBindings as you suggested. Then no users need to be configured in the securityconfig. This is probbaly the cleanest solution.
* Not to reapply the securityconfig if any Users are provisioned. Not ideal as that means the admin user can never be changed. More like a short term solution.

@dbason Do you have any thoughts on this?

**Comment by max-frank** [comment]
I agree using client certs is the cleanest solution here since as you mention re-applying would cause disruptions for any local account configured. Also maybe an added suggestion to that would be support for creating additional client certificates for extra local users when the autogen option is used.

**Comment by swoehrl-mw** [comment]
@anubisg1 That might very well be the best solution. Have the user either use only securityconfig or (basically) only CRDs.

**Comment by saketmht** [comment]
@swoehrl-mw If we want to have a separate CRD for everything (except the config.yml), and remove the same from security-config secret, I would be willing to contribute the same. The security-config overwriting everything created via REST API or Dashboard UI is quite an inconvenience for our use-case as well.

**Comment by arve0** [comment]
> Your suggestion of reapplying the Users after a securityconfig change might work but I think its not a good solution as then there is a time window between securityconfig applied and Users recreated where these users have no access.

The time window will be quite small and happen infrequently?

There will also be a time window when manually editing roles/bindings/users with dashboard/API.

I think that a reconcile that checks current state every say 10-30 minutes and fires after securityconfig-update will be a good enough solution.

**Comment by swoehrl-mw** [comment]
@saketmht Such a PR would be highly appreciated.
For backwards compatibility the old securityconfig logic needs to remain, but it can be changed to only apply the `config.yml` if the others are not defined and apply everything if the ymls are there (that way people can still use securityconfig if they want or they can use CRDs).
One complexity I see: The reconcilers for the users and co currently use the opensearch admin user. Without the securityconfig-job there would be a sort of chicken-egg-problem. The best solution is probably to switch the reconcilers to use a certificate instead.
Another point to consider: The health checks also use the admin-user, so we need a solution for that as well.

**Comment by swoehrl-mw** [comment]
@saketmht 

> Then I will create separate issues (one at a time) for each CRD that I will add. We can discuss the CRD fields/designs in the same issue. Is that okay?

Sounds good. Thank you.

> Do you mean the cluster health via the API?

I'm refering to the kubernetes probes (https://github.com/Opster/opensearch-k8s-operator/blob/main/opensearch-operator/pkg/builders/cluster.go#L199), these are separate from the api calls the operator does.

**Comment by saketmht** [comment]
@swoehrl-mw 
> I'm refering to the kubernetes probes

For this I think, the admin certs (which are used by the securityconfig-job to execute the securityadmin script) can be mounted to the cluster pods as well and pass those certs to the curl command instead of the admin:admin credentials.

For the operator itself, since the operator creates those certs, I don't think we can mount them to the operator pod itself, but we could read the admin-cert secret and use that to create the os client. This way the reconcilers won't need the admin:admin credentials either. I am not sure if its the best way to achieve this though.

**Comment by dbason** [comment]
I think we can also get around this by doing some additional work in the controller to trigger reconciliations when things change (this will require some more advanced usage of controller runtime though)

**Comment by securitym0nkey** [comment]
> @anubisg1 That might very well be the best solution. Have the user either use only securityconfig or (basically) only CRDs.

I was trying exactly this. ** (basically) only CRDs for users, roles, etc** But without the admin in the securityconfig i can't get the cluster to bootstrap.

**Comment by Akila-I** [comment]
> > @anubisg1 That might very well be the best solution. Have the user either use only securityconfig or (basically) only CRDs.
> 
> I was trying exactly this. ** (basically) only CRDs for users, roles, etc** But without the admin in the securityconfig i can't get the cluster to bootstrap.

I experienced the same. Anyone got through this successfully? What can I be missing?

**Comment by synhershko** [comment]
Hey, thanks for opening this and sorry for the wait.

We've made the decision to narrow the operator's scope down to core cluster lifecycle management going forward (provisioning, scaling, upgrades, TLS, etc.) and are deprecating the CRDs that manage OpenSearch-side config like ISM policies, index/component templates, snapshot policies, users, roles, role bindings, tenants and action groups. They're marked deprecated as of #1520 and will be removed in v4.

The reasoning: this is config that's better owned by tools built for declarative OpenSearch config management (e.g. Terraform, or the OpenSearch API/Dashboards directly), and maintaining full parity with these OpenSearch features inside the operator was spreading us too thin on the core stuff people actually depend on.

Given that, we won't be fixing or building out this behavior further, so I'm closing this as won't-fix. If you're relying on this CRD today, it'll keep working until v4, but we'd recommend planning a migration off it. Sorry this isn't the outcome you were hoping for, and thanks again for the report/contribution.
